File: run_simulation.py
from deck import Deck
from blackjack_game import BlackjackGame
import pandas as pd
import streamlit as st

# TASK: add more parameters the user can customize
def run_simulation(num_games, base_bet, initial_bankroll):
    results = {"player_blackjack": 0, "player_bust": 0, "dealer_win": 0, "tie": 0}
    total_wins = {i: 0 for i in range(1, 10)}
    total_ties = 0
    total_winnings = 0
    total_losses = 0
    total_bets = 0

    bankroll = initial_bankroll
    running_count = 0
    bet_sizes = []
    prof = []
    running_counts = []
    bets_vs_running_count = []
    bankroll_over_time = []  

    deck = Deck()

    data = {
        "Total Net Profit": [],
        "Current Game Net Profit": [],
        "Running Count": [],
        "Bet Amount": [],
        "Current Bankroll": [], 
        "Player Win Rate": [],  
        "Dealer Win Rate": [],  
        "Tie Rate": [],  
        "Expected Value Per Game": [],  
        "House Edge": [], 
        "Total Profit Amount": [],  
        "Total Bet Amount": [],  
        "Final Bankroll": []  
    }

    for _ in range(num_games):
        # TASK: Instead of 15 make random int between some to numbers
        if len(deck.deck) <= 15:  # Check if the deck needs to be reshuffled
            deck = Deck()
            running_count = 0

        game = BlackjackGame(deck, running_count)    

        # TASK: give ability to make user create/choose custom bet spread 
        if running_count <= -1:
            bet_amount = 0#base_bet
        elif running_count == 0 or running_count == 1:
            bet_amount = base_bet
        elif running_count == 2:
            bet_amount = base_bet * 4
        elif running_count == 3:
            bet_amount = base_bet * 6
        elif running_count == 4:
            bet_amount = base_bet * 8
        elif running_count == 5:
            bet_amount = base_bet * 12
        elif running_count == 6 or running_count == 7 or running_count == 8 or running_count == 9:
            bet_amount = base_bet * 16
        elif running_count >= 10:
            bet_amount = base_bet * 32
        else:
            bet_amount = base_bet

        # Check if bankroll can cover the bet
        if bankroll < bet_amount:
            st.write("Bankroll depleted! Simulation stopped.")
            break

        bet_sizes.append(bet_amount)
        bets_vs_running_count.append((running_count, bet_amount))  # Track the bet amount and running count

        result = game.start_game()

        total_bets += bet_amount
        curr_game_net_profit = 0

        if result == "player_blackjack":
            results["player_blackjack"] += 1
            total_winnings += bet_amount * 1.5
            curr_game_net_profit += bet_amount * 1.5
        elif "player_win" in result:
            parts = result.split("player_win")
            num_wins = int(parts[0])
            if len(parts) > 1 and parts[1].startswith("_"):
                num_ties = int(parts[1].split("_")[1].split("tie")[0])
                total_ties += num_ties
            total_winnings += bet_amount * num_wins
            curr_game_net_profit += bet_amount * num_wins
            total_wins[num_wins] += 1
        elif result == "dealer_win":
            results["dealer_win"] += 1
            total_losses += bet_amount
            curr_game_net_profit -= bet_amount
        elif result == "player_bust":
            results["player_bust"] += 1
            total_losses += bet_amount
            curr_game_net_profit -= bet_amount
        elif result == "tie":
            results["tie"] += 1

        bankroll += curr_game_net_profit  # Update the bankroll
        bankroll_over_time.append(bankroll)  # Track bankroll over time

        net_profit = total_winnings - total_losses  # Calculate TOTAL net profit
        data["Total Net Profit"].append(net_profit)
        data["Current Game Net Profit"].append(curr_game_net_profit)
        data["Running Count"].append(game.running_count)
        data["Bet Amount"].append(bet_amount)
        data["Current Bankroll"].append(bankroll)  

        running_count = game.running_count
        running_counts.append(running_count)

    player_win_rate = (sum(total_wins.values()) + results["player_blackjack"] + results["tie"] / 2) / num_games
    dealer_win_rate = results["dealer_win"] / num_games
    tie_rate = (results["tie"] + total_ties) / num_games
    expected_value_per_game = net_profit / num_games
    expected_value_per_bet = net_profit / total_bets
    house_edge = -expected_value_per_bet * 100

    # Append summary statistics to the data dictionary
    data["Player Win Rate"].append(player_win_rate * 100)
    data["Dealer Win Rate"].append(dealer_win_rate * 100)
    data["Tie Rate"].append(tie_rate * 100)
    data["Expected Value Per Game"].append(expected_value_per_game)
    data["House Edge"].append(house_edge)
    data["Total Profit Amount"].append(net_profit)
    data["Total Bet Amount"].append(total_bets)
    data["Final Bankroll"].append(bankroll)

    # Ensure all arrays have the same length
    max_length = max(len(value) for value in data.values())
    for key, value in data.items():
        while len(value) < max_length:
            value.append(value[-1])

    return pd.DataFrame(data)

File: blackjack-sim/src/run_simulation.ts
// src/run_simulation.ts

import Deck from "./components/Deck";
import BlackjackGame from "./components/BlackjackGame";
import { DEFAULT_PENETRATION } from "./constants";

interface SimulationData {
  [key: string]: number[];
}

interface Results {
  [key: string]: number;
}

export interface SimulationSummary {
  playerWinRate: number;
  dealerWinRate: number;
  tieRate: number;
  expectedValuePerGame: number;
  houseEdge: number;
  netProfit: number;
  totalBets: number;
  finalBankroll: number;
  numBankruptcies: number;
}

export interface SimulationResult {
  data: SimulationData;
  summary: SimulationSummary;
}

export interface RunSimulationOptions {
  numGames: number;
  baseBet: number;
  initialBankroll: number;
  bettingSpread: Record<string, number>;
  numberOfDecks?: number;
  penetration?: number;
}

interface BettingSpread {
  [key: string]: number;
}

export function trueCountForBetting(
  runningCount: number,
  deck: Deck
): number {
  const decksRemaining = Math.max(0.25, deck.cardsRemaining() / 52);
  return Math.trunc(runningCount / decksRemaining);
}

/** @deprecated Use trueCountForBetting — bet sizing always uses true count. */
export function countForBetting(
  runningCount: number,
  deck: Deck,
  _numberOfDecks?: number,
  _bettingMode?: string
): number {
  return trueCountForBetting(runningCount, deck);
}

export function resolveBetAmount(
  baseBet: number,
  runningCount: number,
  bettingSpread: BettingSpread,
  deck: Deck,
  _numberOfDecks: number = 1
): number {
  const count = trueCountForBetting(runningCount, deck);

  for (const key in bettingSpread) {
    let betMultiplier = 1;
    if (count < 0 && key === "-1") {
      betMultiplier = bettingSpread["-1"];
    } else if (key.includes("-")) {
      const [min, max] = key.split("-").map(Number);
      if (count >= min && count <= max) {
        betMultiplier = bettingSpread[key];
      }
    } else {
      const singleValue = Number(key);
      if (count === singleValue) {
        betMultiplier = bettingSpread[key];
      }
    }
    if (betMultiplier !== 1) {
      return baseBet * betMultiplier;
    }
  }
  return baseBet;
}

function runSimulation(
  numGamesOrOptions: number | RunSimulationOptions,
  baseBet?: number,
  initialBankroll?: number,
  bettingSpread: BettingSpread = {},
  numberOfDecks: number = 1,
  penetration: number = DEFAULT_PENETRATION
): SimulationResult {
  let opts: RunSimulationOptions;
  if (typeof numGamesOrOptions === "object") {
    opts = numGamesOrOptions;
  } else {
    opts = {
      numGames: numGamesOrOptions,
      baseBet: baseBet!,
      initialBankroll: initialBankroll!,
      bettingSpread,
      numberOfDecks,
      penetration,
    };
  }

  const {
    numGames,
    baseBet: bet,
    initialBankroll: bankrollStart,
    bettingSpread: spread,
  } = opts;
  numberOfDecks = opts.numberOfDecks ?? 1;
  penetration = opts.penetration ?? DEFAULT_PENETRATION;

  const results: Results = {
    player_blackjack: 0,
    player_bust: 0,
    dealer_win: 0,
    tie: 0,
  };
  const totalWins: Results = {
    1: 0,
    2: 0,
    3: 0,
    4: 0,
    5: 0,
    6: 0,
    7: 0,
    8: 0,
    9: 0,
  };
  let totalTies = 0;
  let totalWinnings = 0;
  let totalLosses = 0;
  let totalBets = 0;
  let numBankruptcies = 0;

  let bankroll = bankrollStart;
  let runningCount = 0;

  let deck = new Deck(numberOfDecks);
  const clampedPenetration = Math.min(0.95, Math.max(0.5, penetration));
  const reshuffleThreshold = Math.floor(
    deck.initialSize * (1 - clampedPenetration)
  );

  const data: SimulationData = {
    "Total Net Profit": [],
    "Current Game Net Profit": [],
    "Running Count": [],
    "Bet Amount": [],
    "Current Bankroll": [],
    "Player Win Rate": [],
    "Dealer Win Rate": [],
    "Tie Rate": [],
    "Expected Value Per Game": [],
    "House Edge": [],
    "Total Profit Amount": [],
    "Total Bet Amount": [],
    "Final Bankroll": [],
    "Number of Bankruptcies": [],
  };

  let gamesPlayed = 0;

  for (let i = 0; i < numGames; i++) {
    if (deck.cardsRemaining() <= reshuffleThreshold) {
      deck = new Deck(numberOfDecks);
      runningCount = 0;
    }

    const betAmount = spread
      ? resolveBetAmount(bet, runningCount, spread, deck, numberOfDecks)
      : bet;

    if (bankroll < betAmount) {
      numBankruptcies += 1;
      break;
    }

    const game = new BlackjackGame(deck, runningCount);
    const gameResult = game.startGame(betAmount);

    totalBets += gameResult.totalWagered;
    const currGameNetProfit = gameResult.totalProfit;
    totalWinnings += Math.max(0, currGameNetProfit);
    totalLosses += Math.max(0, -currGameNetProfit);

    if (gameResult.playerBlackjack) results.player_blackjack += 1;
    if (gameResult.playerBust) results.player_bust += 1;
    if (gameResult.dealerWin && gameResult.winCount === 0) {
      results.dealer_win += 1;
    }
    if (
      gameResult.tieCount === gameResult.hands.length &&
      gameResult.winCount === 0
    ) {
      results.tie += 1;
    }
    if (gameResult.winCount > 0) {
      totalWins[gameResult.winCount] =
        (totalWins[gameResult.winCount] ?? 0) + 1;
    }
    totalTies += gameResult.tieCount;

    bankroll += currGameNetProfit;
    gamesPlayed += 1;

    const netProfit = bankroll - bankrollStart;
    data["Total Net Profit"].push(netProfit);
    data["Current Game Net Profit"].push(currGameNetProfit);
    data["Running Count"].push(gameResult.runningCount);
    data["Bet Amount"].push(gameResult.totalWagered);
    data["Current Bankroll"].push(bankroll);

    runningCount = gameResult.runningCount;
  }

  const denom = gamesPlayed || 1;
  const playerWinRate =
    (Object.values(totalWins).reduce((a, b) => a + b, 0) +
      results.player_blackjack +
      results.tie / 2) /
    denom;
  const dealerWinRate = results.dealer_win / denom;
  const tieRate = (results.tie + totalTies) / denom;
  const netProfit = bankroll - bankrollStart;
  const expectedValuePerGame = netProfit / denom;
  const expectedValuePerBet = totalBets === 0 ? 0 : netProfit / totalBets;
  const houseEdge = -expectedValuePerBet * 100;

  data["Player Win Rate"].push(playerWinRate * 100);
  data["Dealer Win Rate"].push(dealerWinRate * 100);
  data["Tie Rate"].push(tieRate * 100);
  data["Expected Value Per Game"].push(expectedValuePerGame);
  data["House Edge"].push(houseEdge);
  data["Total Profit Amount"].push(netProfit);
  data["Total Bet Amount"].push(totalBets);
  data["Final Bankroll"].push(bankroll);
  data["Number of Bankruptcies"].push(numBankruptcies);

  const maxLength = Math.max(...Object.values(data).map((arr) => arr.length));
  for (const key in data) {
    while (data[key].length < maxLength) {
      data[key].push(data[key][data[key].length - 1]);
    }
  }

  return {
    data,
    summary: {
      playerWinRate,
      dealerWinRate,
      tieRate,
      expectedValuePerGame,
      houseEdge,
      netProfit,
      totalBets,
      finalBankroll: bankroll,
      numBankruptcies,
    },
  };
}

export default runSimulation;
